rskj-core/src/main/java/co/rsk/net/sync/DownloadingBodiesSyncState.java

```diff
@@ -1,69 +1,301 @@
+            syncEventsHandler.onCompletedSyncing();
+        }
+    }
+
+    private void tryRequestNextBody(NodeID peerId) {
+        Optional<BlockHeader> nextHeader = updateHeadersAndChunks(peerId, chunksBeingDownloaded.get(peerId));
+        if (nextHeader.isPresent()){
+            requestBody(peerId, nextHeader.get());
+        }
+    }
+
+    private void handleInvalidMessage(NodeID peerId, BlockHeader header) {
+        syncInformation.reportEvent(
+                "Invalid body received from node {}",
+                EventType.INVALID_MESSAGE, peerId);
 
-        if (!pendingHeaders.isEmpty()) {
-            requestBody();
+        clearPeerInfo(peerId);
+        if (suitablePeers.isEmpty()){
+            syncEventsHandler.stopSyncing();
             return;
         }
+        messagesByPeers.remove(peerId);
+        resetChunkAndHeader(peerId, header);
+        startDownloading(getInactivePeers());
+    }
+
+    private void handleUnexpectedBody(NodeID peerId) {
+        syncInformation.reportEvent(
+                "Unexpected body received from node {}",
+                EventType.UNEXPECTED_MESSAGE, peerId);
 
-        // Finished syncing
-        syncEventsHandler.onCompletedSyncing();
+        clearPeerInfo(peerId);
+        if (suitablePeers.isEmpty()) {
+            syncEventsHandler.stopSyncing();
+            return;
+        }
+        // if this peer has another different message pending then its restored to the stack
+        Long messageId = messagesByPeers.remove(peerId);
+        if (messageId != null) {
+            resetChunkAndHeader(peerId, pendingBodyResponses.remove(messageId).header);
+        }
+        startDownloading(getInactivePeers());
+    }
+
+    private void resetChunkAndHeader(NodeID peerId, BlockHeader header) {
+        Integer chunkNumber = chunksBeingDownloaded.remove(peerId);
+        pendingHeaders.get(chunkNumber).push(header);
+        Integer segmentNumber = segmentsBeingDownloaded.remove(peerId);
+        chunksBySegment.get(segmentNumber).push(chunkNumber);
+    }
+
+    private void clearPeerInfo(NodeID peerId) {
+        suitablePeers.remove(peerId);
+        timeElapsedByPeer.remove(peerId);
+    }
+
+    private Optional<BlockHeader> updateHeadersAndChunks(NodeID peerId, Integer currentChunk) {
+        Stack<BlockHeader> headers = pendingHeaders.get(currentChunk);
+        if (!headers.empty()) {
+            return Optional.of(headers.pop());
+        }
+
+        Optional<BlockHeader> header = tryFindBlockHeader(peerId);
+        if (!header.isPresent()){
+            chunksBeingDownloaded.remove(peerId);
+            segmentsBeingDownloaded.remove(peerId);
+            messagesByPeers.remove(peerId);
+        }
+
+        return header;
+    }
+
+    private Optional<BlockHeader> tryFindBlockHeader(NodeID peerId) {
+        Integer segmentNumber = segmentByNode.get(peerId);
+        // we start from the last chunk that can be downloaded
+        while (segmentNumber >= 0){
```

Comment:
a `while` statement with a defined number of iterations should be changed to a `for` loop.

Reply:
:+1: